bindings/go/ctf/artifact_set.go

```diff
@@ -0,0 +1,333 @@
+//	  mediaType: application/vnd.oci.image.manifest.v1+tar+gzip
+//	  referenceName: jakobmoellerdev/podinfo/podinfo:6.7.1
+//	  type: localBlob
+//
+// This ArtifactSetMediaType can be used on artifact introspection on the resource
+// together with NewArtifactSetFromBlob to access that ArtifactSet.
+const ArtifactSetMediaType = "application/vnd.oci.image.manifest.v1+tar+gzip"
+
+// ArtifactSet is the equivalent (deprecated) implementation of
+// https://github.com/open-component-model/ocm/blob/main/api/oci/extensions/repositories/artifactset
+//
+// NOTE: Even though it does look similar to an OCI Image Layout, it is not
+// the same. Notable differences are:
+//
+//   - The blobs directory orders blobs in the form
+//
+//     blobs/digest-algo.digest
+//
+//     instead of
+//
+//     blobs/digest-algo/digest.
+//
+//   - The index.json file IS a valid OCI Image Index json, but does not maintain
+//     the ociimagespec.AnnotationRefName correctly, instead it sets it to the version
+//     of the resource. Additionally, it maintains the software.ocm/tags annotation.
+//     This means it is MANDATORY to understand the resource access.referenceName
+//     in place of the usual meaning of the refName to fully target an OCI Image out
+//     of an ArtifactSet.
+//
+//   - The index.json file contains a software.ocm/main annotation, that declares
+//     the main blob to be introspected (useful for single-layer artifacts or multi-layer
+//     artifacts with one main layer and multiple metadata layers).
+//
+// Altogether, this makes the ArtifactSet a custom format that is not compatible
+// with OCI Image Layouts or CTF readings that are unaware of the Component Descriptor.
+//
+// It is thus deprecated and should not be used anymore.
+//
+// This ArtifactSetMediaType now only serves to read from old LocalBlobs and CTFs and is maintained
+// to ensure compatibility.
+//
+// New CTFs should and will always include localBlobs not as artifact sets, but as proper
+// OCI Image Layouts.
+type ArtifactSet interface {
+	io.Closer
+	ReadOnlyBlobStore
+	GetIndex() ociimagespec.Index
+}
+
+var _ ArtifactSet = (*artifactSet)(nil)
+
+// NewArtifactSetFromBlob creates a new ArtifactSet from a blob.
+// It will start owning the readcloser in the blob and needs to be closed due to this.
+// It is backed by a tar filesystem that is either read from a tar file or a gzip file.
+// The gzip format is not detected by ArtifactSetMediaType, but by the magic number in the first 512 bytes of the file.
+// The blob must be a valid ArtifactSet, otherwise an error is returned.
+//
+// The ArtifactSet mereley supports read-only and any future write operation is no longer supported.
+// All ArtifactSet's encountered in the wild MUST be converted to OCI Image Layouts.
+//
+// See ArtifactSet for more information about the format.
+func NewArtifactSetFromBlob(b blob.ReadOnlyBlob) (ArtifactSet, error) {
```

Comment:
Do we really want to return an interface here?

Reply:
I see why. Any reason to have it backed by an unexported type? This will make this a bit difficult to work with from the outside.

Reply:
Can expose the type directly too since all fields are unexported. 👍 I think generally it should not make a difference right now